Add TpetraWrapper::BlockVector and TpetraWrappers::BlockSparseMatrix

kinnewig opened the pull request:
This PR adds support for block systems to the TpetraWrappers.
The `TpetraWrappers::BlockVector` class is mainly based on the `TrilinosWrappers:MPI::BlockVector` class. Analogous, the `TpetraWrappers::BlockSparseMatrix` class is primarily based on the `TrilinosWrappers::BlockSparseMatrix`.

For the moment, I separated this PR into several commits. If it is preferred, I can squash it into one.

bangerth reviewed (commented):
Is it correct that there is really nothing here that is interesting? I would think that you probably just copied the existing Epetra block matrix thing and changed type names where necessary, but other than that the code is exactly as for the Epetra version of the class?

masterleinad reviewed (commented):
Can we enable some unit tests?

kinnewig commented:
> Is it correct that there is really nothing here that is interesting? I would think that you probably just copied the existing Epetra block matrix thing and changed type names where necessary, but other than that the code is exactly as for the Epetra version of the class?

Yes, with a few exceptions, this is mainly copied from the Epetra block class, and I adapted the names. A few functions that were hard to translate from Epetra to Tpetra (and not necessary for my case) are left out.
Therefore, I don't know if this should be marked as a major change.

kinnewig commented on include/deal.II/lac/trilinos_tpetra_block_sparse_matrix.templates.h:
```diff
@@ -0,0 +1,469 @@
+
+
+
+    // template <typename Number, typename MemorySpace>
+    // template<>
+    // void
+    // BlockSparseMatrix<Number, MemorySpace>::reinit(const BlockSparsityPattern
+    // &block_sparsity_pattern)
+    //{
+    //   // Call the other basic reinit function, ...
+    //   reinit(block_sparsity_pattern.n_block_rows(),
+    //          block_sparsity_pattern.n_block_cols());
+
+    //  // ... set the correct sizes, ...
+    //  this->row_block_indices    = block_sparsity_pattern.get_row_indices();
+    //  this->column_block_indices =
+    //  block_sparsity_pattern.get_column_indices();
+
+    //  // ... and then assign the correct
+    //  // data to the blocks.
+    //  for (size_type r = 0; r < this->n_block_rows(); ++r)
+    //    for (size_type c = 0; c < this->n_block_cols(); ++c)
+    //      {
+    //        this->sub_objects[r][c]->reinit(block_sparsity_pattern.block(r,
+    //        c));
+    //      }
+    //}
```
I just skipped over this function since it was not necessary for a working set and then forgot about it. 
I have another look into this function. If there is no straight-up way to implement it, I will delete it.

kinnewig commented:
> Can we enable some unit tests?

Sure! Especially since copying the test cases from the Epetra case should be possible.

bangerth commented:
/rebuild

bangerth commented:
@masterleinad Were you hoping for tests as part of this PR, or for future PRs?

masterleinad commented:
> @masterleinad Were you hoping for tests as part of this PR, or for future PRs?

I was hoping to have some tests here.

kinnewig commented:
I have already continued the work on this PR. 
As a first test, I tried to solve step-31 using the  Tpetra Block System. During that test, I found a lot of bugs. And I am still working on fixing all those bugs. 

I will resolve the merge conflict once I have fixed all the bugs.

bangerth commented:
Ah, awesome!

kinnewig commented:
I finally found the time to continue working on this PR.
I removed a test that failed because it requires vmult between a dealii::vector and a TpetraWrappers::BlockMatrix (and I moved the vmult into a separate PR).

I removed the changelog entry. I like the idea of creating one changelog entry for the complete Tpetra-Interface.

And I resolved a merge conflict.

From my side, this PR is ready to be merged. So, if there are no further remarks, I would squash the commitments into a reasonable amount.

kinnewig commented on doc/news/changes/major/20240201SebastianKinnewig:
```diff
@@ -0,0 +1,8 @@
+based on LinearAlgebra::TpetraWrappers::Vector and 
+LinearAlgebra::TpetraWrappers::SparseMatrix.
+<br>
+(Sebastian Kinnewig, 2023/11/28)
```
I like that idea, let's do that!

I removed the changelog entry from this PR.

bangerth reviewed (commented):
This looks good, but could I ask you to rearrange the 11 commits in a way that each is somewhat self-contained and complete? It doesn't have to be less than 11 commits, given that you touch several of the existing classes, but perhaps we can make them all stand on their own, instead of some commits fixing previous commits.

Separately, are the tests you add identical to the ones in #16611? If so, could they go into the `trilinos_tpetra/` directory?

kinnewig commented:
> This looks good, but could I ask you to rearrange the 11 commits in a way that each is somewhat self-contained and complete? It doesn't have to be less than 11 commits, given that you touch several of the existing classes, but perhaps we can make them all stand on their own, instead of some commits fixing previous commits.

Sure! And done!

> Separately, are the tests you add identical to the ones in #16611? If so, could they go into the `trilinos_tpetra/` directory?

Done!

kinnewig commented:
I am not sure why `indent` failed.

The error message occurs in the documentation step:
```
error: could not copy file /home/runner/work/dealii/dealii/doc/doxygen/images/dgp_doesnt_contain_p.png to /home/runner/work/dealii/dealii/build/doc/doxygen/./deal.II/dgp_doesnt_contain_p.png
```

bangerth commented:
The tests fail because of this linker error:
```
/__w/dealii/dealii/build/lib/libdeal_II.g.so.9.6.0-pre: error: undefined reference to 'void dealii::LinearAlgebra::TpetraWrappers::SparseMatrix<double, dealii::MemorySpace::Host>::reinit<dealii::SparseMatrix<double> >(dealii::IndexSet const&, dealii::IndexSet const&, dealii::SparseMatrix<double> const&, ompi_communicator_t*, bool)'
```

kinnewig commented:
> The tests fail because of this linker error:
> 
> ```
> /__w/dealii/dealii/build/lib/libdeal_II.g.so.9.6.0-pre: error: undefined reference to 'void dealii::LinearAlgebra::TpetraWrappers::SparseMatrix<double, dealii::MemorySpace::Host>::reinit<dealii::SparseMatrix<double> >(dealii::IndexSet const&, dealii::IndexSet const&, dealii::SparseMatrix<double> const&, ompi_communicator_t*, bool)'
> ```

I am not aware that I modified anything related to the function
```
SparseMatrix<Number, MemorySpace>::reinit(
  const IndexSet &,
  const IndexSet &,
  dealii::SparseMatrix<Number>,
  ompi_communicator_t*,
  bool)
```

Moreover, I can't spot this function in the master branch...
Does this test even work?

bangerth commented:
It's this function:
```
      template <typename SparsityPatternType>
      void
      reinit(const IndexSet            &row_parallel_partitioning,
             const IndexSet            &col_parallel_partitioning,
             const SparsityPatternType &sparsity_pattern,
             const MPI_Comm             communicator  = MPI_COMM_WORLD,
             const bool                 exchange_data = false);
```
It expects a sparsity pattern as argument, but it is called with a sparse matrix. I'll look around where that is.

bangerth commented on include/deal.II/lac/trilinos_tpetra_block_sparse_matrix.templates.h:
```diff
@@ -0,0 +1,340 @@
+
+    template <typename Number, typename MemorySpace>
+    void
+    BlockSparseMatrix<Number, MemorySpace>::reinit(
+      const std::vector<IndexSet>               &parallel_partitioning,
+      const ::dealii::BlockSparseMatrix<double> &dealii_block_sparse_matrix,
+      const MPI_Comm                             communicator,
+      const double                               drop_tolerance)
+    {
+      const size_type n_block_rows = parallel_partitioning.size();
+
+      Assert(n_block_rows == dealii_block_sparse_matrix.n_block_rows(),
+             ExcDimensionMismatch(n_block_rows,
+                                  dealii_block_sparse_matrix.n_block_rows()));
+      Assert(n_block_rows == dealii_block_sparse_matrix.n_block_cols(),
+             ExcDimensionMismatch(n_block_rows,
+                                  dealii_block_sparse_matrix.n_block_cols()));
+
+      // Call the other basic reinit function ...
+      reinit(n_block_rows, n_block_rows);
+
+      // ... and then assign the correct
+      // data to the blocks.
+      for (size_type r = 0; r < this->n_block_rows(); ++r)
+        for (size_type c = 0; c < this->n_block_cols(); ++c)
+          {
+            this->sub_objects[r][c]->reinit(parallel_partitioning[r],
+                                            parallel_partitioning[c],
+                                            dealii_block_sparse_matrix.block(r,
+                                                                             c),
+                                            communicator,
+                                            drop_tolerance);
```
I think it is this place. The function takes a block sparse matrix as argument, and so here we call `reinit()` with a sparse matrix as third argument, but the function expects a sparsity pattern.

I think what is necessary is to provide an overload of `LA::TpetraWrappers::SparseMatrix::reinit (..., dealii::SparseMatrix, ...).

bangerth commented:
For reference, in the Epetra wrappers, the `reinit()` function question is declared as follows:
```
    template <typename SparsityPatternType>
    std::enable_if_t<
      !std::is_same_v<SparsityPatternType, dealii::SparseMatrix<double>>>
    reinit(const IndexSet            &row_parallel_partitioning,
           const IndexSet            &col_parallel_partitioning,
           const SparsityPatternType &sparsity_pattern,
           const MPI_Comm             communicator  = MPI_COMM_WORLD,
           const bool                 exchange_data = false);
```
Apparently someone had fallen into the same trap before, and so the `std::enable_if` was added. I will add it to the Tpetra wrappers too in a separate PR.

kinnewig commented:
Thank you very much for the detailed feedback!

bangerth commented:
If you haven't already found all places where this happens, rebase on master now that #16654 has been merged and the compiler will tell you.

kinnewig commented:
> If you haven't already found all places where this happens, rebase on master now that #16654 has been merged and the compiler will tell you.

Thank you, again!
With the rebase, the compiler notices that some functions are not present.

kinnewig commented:
I added the missing function.
First, I had to rework the `reinit_matrix()` function from LA::TpetraWrappers::SparseMatrix. I changed the original `reinit_matrix()`  to be a specialization for DynamicSparsityPattern and added a new templated version of `reinit_matrix()`.
This is included in commit [ec280f3](https://github.com/dealii/dealii/pull/16588/commits/ec280f30600e81953ff12ee32aca46094a4ec77c)

Secondly, I added the missing reinit() function in commit [a989f0c](https://github.com/dealii/dealii/pull/16588/commits/a989f0c63e7851b3ae149f146f95a51f2a6d1238).

I am quite confident that this should work now.

bangerth commented:
It currently still fails with this error:
```
/__w/dealii/dealii/include/deal.II/lac/trilinos_tpetra_sparse_matrix.templates.h:626:46: error: ‘class Tpetra::CrsMatrix<double, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >’ has no member named ‘getLocalMaxNumRowEntries’; did you mean ‘getGlobalMaxNumRowEntries’?
  626 |       size_type maximum_row_length = matrix->getLocalMaxNumRowEntries();
      |                                      ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
      |                                      getGlobalMaxNumRowEntries
```
Is that an issue with different Trilinos versions perhaps?

kinnewig commented:
> It currently still fails with this error:
> 
> ```
> /__w/dealii/dealii/include/deal.II/lac/trilinos_tpetra_sparse_matrix.templates.h:626:46: error: ‘class Tpetra::CrsMatrix<double, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >’ has no member named ‘getLocalMaxNumRowEntries’; did you mean ‘getGlobalMaxNumRowEntries’?
>   626 |       size_type maximum_row_length = matrix->getLocalMaxNumRowEntries();
>       |                                      ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
>       |                                      getGlobalMaxNumRowEntries
> ```
> 
> Is that an issue with different Trilinos versions perhaps?

Yes, it is a problem with different Trilinos versions. Actually, I have encountered that problem already at some other places, too.
I will push a fix for that in a moment
